autogen/oai/client.py

```diff
@@ -30,12 +33,215 @@
+        elif context:
+            # Instantiate the messages
+            params["messages"] = [
+                {
+                    **m,
+                    "content": self.instantiate(m["content"], context, allow_format_str_template),
+                }
+                if m.get("content")
+                else m
+                for m in messages
+            ]
+        return params
+
+    def create(self, client, client_id, is_last, create_config: Dict, extra_kwargs: Dict):
+        # construct the create params
+        params = self.construct_create_params(create_config, extra_kwargs)
+        # get the cache_seed, filter_func and context
+        cache_seed = extra_kwargs.get("cache_seed", 41)
+        filter_func = extra_kwargs.get("filter_func")
+        context = extra_kwargs.get("context")
+        with diskcache.Cache(f"{self.cache_path_root}/{cache_seed}") as cache:
+            if cache_seed is not None:
+                # Try to get the response from cache
+                key = get_key(params)
+                response = cache.get(key, None)
+                if response is not None:
+                    # check the filter
+                    pass_filter = filter_func is None or filter_func(context=context, response=response)
+                    if pass_filter or is_last:
+                        # Return the response if it passes the filter or it is the last client
+                        response.pass_filter = pass_filter
+                        # TODO: add response.cost
+                        return response
+
+            response = client.create(params)
+            if response is None:
+                return None
+
+            if cache_seed is not None:
+                # Cache the response
+                with diskcache.Cache(f"{self.cache_path_root}/{cache_seed}") as cache:
+                    cache.set(key, response)
+
+            # check the filter
+            pass_filter = filter_func is None or filter_func(context=context, response=response)
+            if pass_filter or is_last:
+                # Return the response if it passes the filter or it is the last client
+                response.pass_filter = pass_filter
+                response.config_id = client_id
+                response.cost = client.cost(response)
+                return response
+            return None
+
+
+class Client(ABC):
+    """
+    A client class must implement the following methods:
+    - create must return a response object that implements the ClientResponseProtocol
+    - cost
+
+    This class is used to create a client that can be used by OpenAIWrapper.
+    It mimicks the OpenAI class, but allows for custom clients to be used.
+    """
+
+    class ClientResponseProtocol(Protocol):
+        class Choice(Protocol):
+            class Message(Protocol):
+                content: str | None
+                function_call: str | None
```

Comment:
The new openai v1.3 might break this protocol. See #791

Reply:
added an extra check to ignore anything chat completions related during extraction

Ideally this stabilizes to a universal API that can be re-used by different client implementations